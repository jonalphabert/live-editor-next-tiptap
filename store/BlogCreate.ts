import { DraftStatus } from '@/models/DraftStatus';
import TurndownService from 'turndown';
import { create } from 'zustand';

type BlogCreateState = {
  blogPostId: string | null;
  title: string;
  setTitle: (title: string) => void;
  content: string;
  setContent: (content: string) => void;
  isOpen: boolean;
  openModal: () => void;
  closeModal: () => void;
  description: string;
  setDescription: (desc: string) => void;
  selectedGroup: string | null;
  setSelectedGroup: (group: string) => void;
  selectedCategories: string[];
  isEdited: boolean;
  toggleCategory: (category: string) => void;
  resetForm: () => void;
  submitForm: () => Promise<void>;
  draftPost: () => Promise<DraftStatus>;
  saveAsMarkdown: () => void;
};

export const useModalStore = create<BlogCreateState>((set: Function, get: Function) => ({
  blogPostId: null,
  isOpen: false,
  description: '',
  title: '',
  content: 'Start typing here...',
  isEdited: false,
  selectedGroup: null,
  selectedCategories: [],
  
  openModal: () => set({ isOpen: true }),
  closeModal: () => set({ isOpen: false }),

  setTitle: (title: string) => {
    set({ title: title });

    if(get().isEdited) return;
    set({ isEdited: true });
  },

  setContent: (content: string) => {
    set({ content: content });

    if(get().isEdited) return;
    set({ isEdited: true });
  },
  
  setDescription: (desc: string) => set({ description: desc }),
  
  setSelectedGroup: (group: string) => set({ selectedGroup: group }),
  
  toggleCategory: (category : string) => set((state : BlogCreateState) => ({
    selectedCategories: state.selectedCategories.includes(category)
      ? state.selectedCategories.filter((c : string) => c !== category)
      : [...state.selectedCategories, category]
  })),
  
  resetForm: () => set({
    description: '',
    selectedGroup: null,
    selectedCategories: []
  }),
  
  submitForm: async () => {
    const { description, selectedGroup, selectedCategories, title, content } = get();
    // API call would go here
    console.log('Submitting:', { description, selectedGroup, selectedCategories, title, content });
    get().closeModal();
    get().resetForm();
  },

  draftPost: async() => {
    const { title, content, blogPostId } = get();
    if(title === '' && content === '') return {
      success: false,
      message: 'Please enter a title and content first'
    };

    if(blogPostId){
      // Hit update blog API
    } else {
      // Hit create blog API
    }

    console.log('Submitting:', { title, content });
    return {
      success: true,
      message: 'Success'
    }
  },
  saveAsMarkdown: () => {
    const { content } = get();
    const turndownService = new TurndownService();
    
    // Add custom rules for better formatting (optional)
    turndownService.addRule('preformatted', {
      filter: ['pre'],
      replacement: (content) => `\`\`\`\n${content}\n\`\`\``
    });
    
    try {
      const markdown = turndownService.turndown(content);
      
      // Create download
      const blob = new Blob([markdown], { type: 'text/markdown' });
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.href = url;
      a.download = 'Readme.md';
      document.body.appendChild(a);
      a.click();
      
      // Cleanup
      setTimeout(() => {
        document.body.removeChild(a);
        URL.revokeObjectURL(url);
      }, 100);
    } catch (error) {
      console.error('Conversion failed:', error);
      alert('Error converting content to Markdown');
    }
  }
}));