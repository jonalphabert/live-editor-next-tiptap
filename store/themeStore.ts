import { create } from 'zustand';
import { persist, createJSONStorage } from 'zustand/middleware';

type ThemeState = {
  darkMode: boolean;
  toggleDarkMode: () => void;
};

// Helper function to detect system theme
const getSystemThemePreference = (): boolean => {
  // Check for SSR (window won't be available)
  if (typeof window === 'undefined') return false;
  
  return window.matchMedia('(prefers-color-scheme: dark)').matches;
};

export const useThemeStore = create<ThemeState>()(
  persist(
    (set) => ({
      darkMode: getSystemThemePreference(),
      toggleDarkMode: () => set((state) => ({ darkMode: !state.darkMode })),
    }),
    {
      name: 'theme-storage',
      storage: createJSONStorage(() => localStorage),
      // Only rehydrate (load from storage) if the storage exists
      partialize: (state) => ({ darkMode: state.darkMode }),
    }
  )
);