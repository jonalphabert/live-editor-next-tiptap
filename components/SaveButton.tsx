"use client"
import { useModalStore } from "@/store/BlogCreate";
import { Button } from "@/components/ui/button";
import { Save } from "lucide-react";

const SaveButton: React.FC = () => {
  const saveAsMarkdown = useModalStore((state) => state.saveAsMarkdown);

  return (
    <Button onClick={saveAsMarkdown}>
      <Save /> Save
    </Button>
  );
};

export default SaveButton;
