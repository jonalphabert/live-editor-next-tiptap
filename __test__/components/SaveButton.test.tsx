import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import SaveButton from "@/components/SaveButton";
import { useModalStore } from "@/store/BlogCreate";

jest.mock('@/store/BlogCreate', () => ({
    useModalStore: jest.fn(),
  }));

describe("SaveButton", () => {
    const mockSaveAsMarkdown = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        (useModalStore as unknown as jest.Mock).mockImplementation((selector) => {
          return selector({ saveAsMarkdown: mockSaveAsMarkdown });
        });
      });

    it("renders correctly", () => {
        render(<SaveButton />);
        expect(screen.getByText("Save")).toBeInTheDocument();
    });

    it('calls saveAsMarkdown function when clicked', () => {
        render(<SaveButton />);
        
        const button = screen.getByRole('button', { name: /save/i });
        fireEvent.click(button);
        
        expect(mockSaveAsMarkdown).toHaveBeenCalledTimes(1);
      });

    it('matches snapshot', () => {
        const { asFragment } = render(<SaveButton />);
        expect(asFragment()).toMatchSnapshot();
      });
});