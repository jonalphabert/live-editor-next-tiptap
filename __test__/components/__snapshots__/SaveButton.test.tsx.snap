// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SaveButton Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-save"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.2 3a2 2 0 0 1 1.4.6l3.8 3.8a2 2 0 0 1 .6 1.4V19a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2z"
      />
      <path
        d="M17 21v-7a1 1 0 0 0-1-1H8a1 1 0 0 0-1 1v7"
      />
      <path
        d="M7 3v4a1 1 0 0 0 1 1h7"
      />
    </svg>
     Save
  </button>
</DocumentFragment>
`;

exports[`SaveButton matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-save"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.2 3a2 2 0 0 1 1.4.6l3.8 3.8a2 2 0 0 1 .6 1.4V19a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2z"
      />
      <path
        d="M17 21v-7a1 1 0 0 0-1-1H8a1 1 0 0 0-1 1v7"
      />
      <path
        d="M7 3v4a1 1 0 0 0 1 1h7"
      />
    </svg>
     Save
  </button>
</DocumentFragment>
`;
